src/domain/entities/DataSet.ts

```diff
@@ -141,6 +144,77 @@ export class DataSet extends Struct<DataSetAttrs>() {
               ];
     }
 
+    validateIndicatorMatching(): ValidationError<DataSet>[] {
+        const indicatorMap = _(this.indicators)
+            .filter(indicator => indicator.type === matchingIndicatorType)
+            .keyBy(indicator => indicator.id);
+        return _(this.indicatorMatching || [])
+            .map(({ source, target }) => [
+                this.validateIndicatorMatchingByRole(source, "source", indicatorMap),
+                this.validateIndicatorMatchingByRole(target, "target", indicatorMap),
+                this.validateIndicatorMatchingCategoryCombo(source, indicatorMap),
+            ])
+            .flatten()
+            .compact()
+            .value();
+    }
+
+    private validateIndicatorMatchingByRole(
+        id: Id,
+        role: "target" | "source",
+        indicatorMap: HashMap<Id, Indicator>
+    ): ValidationError<DataSet> | undefined {
```

Comment:
Nothing wrong with this, but can we use the Maybe<T> type? Just to be consistent with the rest of the code.